layout-server/src/main/java/org/finos/vuu/layoutserver/service/ApplicationLayoutService.java

```diff
@@ -0,0 +1,71 @@
+package org.finos.vuu.layoutserver.service;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.RequiredArgsConstructor;
+import org.finos.vuu.layoutserver.model.ApplicationLayout;
+import org.finos.vuu.layoutserver.repository.ApplicationLayoutRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+
+@RequiredArgsConstructor
+@Service
+public class ApplicationLayoutService {
+
+    private static final Logger logger = LoggerFactory.getLogger(ApplicationLayoutService.class);
+    private static final String DEFAULT_LAYOUT_FILE = "defaultLayout.json";
+    private static ApplicationLayout defaultLayout;
+    private final ApplicationLayoutRepository repository;
+
+    public void createApplicationLayout(String username, JsonNode layoutDefinition) {
+        repository.save(new ApplicationLayout(username, layoutDefinition));
+    }
+
+    public ApplicationLayout getApplicationLayout(String username) {
+        return repository.findById(username).orElseGet(() -> {
+            logger.info("No application layout for user, returning default");
+            return getDefaultLayout();
+        });
+    }
+
+    public void updateApplicationLayout(String username, JsonNode layoutDefinition) {
+        createApplicationLayout(username, layoutDefinition);
+    }
+
+    public void deleteApplicationLayout(String username) {
+        try {
+            repository.deleteById(username);
+        } catch (EmptyResultDataAccessException e) {
+            throw new NoSuchElementException("No layout found for user: " + username);
+        }
+    }
+
+    private ApplicationLayout getDefaultLayout() {
+        if (defaultLayout == null) {
+            loadDefaultLayout();
+        }
+        return defaultLayout;
+    }
+
+    private void loadDefaultLayout() {
+        JsonNode definition = loadDefaultLayoutJsonFile();
+        defaultLayout = new ApplicationLayout(null, definition);
+    }
+
+    private JsonNode loadDefaultLayoutJsonFile() {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            ClassPathResource resource = new ClassPathResource(DEFAULT_LAYOUT_FILE);
+            return objectMapper.readTree(resource.getInputStream());
+        } catch (IOException e) {
+            logger.warn("Failed to read default application layout, returning empty node");
+            return objectMapper.createObjectNode();
```

Comment:
That's correct. Perhaps it would be better to throw the exception and return a 500 though. What do you think?

Reply:
I agree with a 500 because if neither a user's layout nor the default layout is retrieved, something has gone wrong with the server.

Reply:
agreed with the 500 approach. Also I am not sure what would happen if we tried and loaded an empty JSON as an application layout on the client side. It would probably result in some "tried to read property * of undefined" error

Reply:
I'm now throwing a 500 if the file fails to load. I've made some changes to the error handling to accomodate this, and have done some refactor to allow a sensible integration test (`getApplicationLayout_defaultFailsToLoad_returns500`).